homeassistant/components/leviosa_shades/config_flow.py

```diff
@@ -0,0 +1,156 @@
+"""Config flow for Leviosa shades Zone."""
+import logging
+
+from leviosapy import discover_leviosa_zones, validate_zone
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST, CONF_NAME
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import (
+    BLIND_GROUPS,
+    DEVICE_FW_V,
+    DEVICE_MAC,
+    DOMAIN,
+    GROUP1_NAME,
+    GROUP2_NAME,
+    GROUP3_NAME,
+    GROUP4_NAME,
+    GROUP5_NAME,
+    GROUP6_NAME,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME): str,
+        vol.Required(GROUP1_NAME): str,
+        vol.Optional(GROUP2_NAME): str,
+        vol.Optional(GROUP3_NAME): str,
+        vol.Optional(GROUP4_NAME): str,
+        vol.Optional(GROUP5_NAME): str,
+        vol.Optional(GROUP6_NAME): str,
+    }
+)
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class LeviosaConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Manages the interaction with user when a Leviosa Zone needs to be setup."""
+
+    # The schema version below will be used by Home Assistant to determine
+    # if a call to the migrate method is needed; this is not implemented
+    # as of March 2021
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+    GROUPS = [
+        GROUP1_NAME,
+        GROUP2_NAME,
+        GROUP3_NAME,
+        GROUP4_NAME,
+        GROUP5_NAME,
+        GROUP6_NAME,
+    ]
+
+    def __init__(self):
+        """Initialize the Motion Blinds flow."""
+
+        self._host = None
+        self._host_uid = None
+        self._devices = {}
+
+    async def async_step_user(self, user_input=None):
+        """Perform discovery and present an input screen for each Zone discovered."""
+
+        _LOGGER.debug("Looking for Leviosa Zone HUBs")
+        self._devices = await discover_leviosa_zones()
```

Comment:
What kind of discovery is this? Normally we should not use the user flow for discovery, but the existing SSDP / DHCP discovery.

Reply:
We use the SSDP protocol for device discovery, and I know HA already supports SSDP-based auto-discovery, but I still prefer to put it inside leviosa_shades instead of reusing the public SSDP discovery functionality because I want to customize the presentation of the device discovery page in a way that can help users understand the difference between different leviosa shades, and I don't want to scatter discovered leviosa shades in a massive list of chanting HA auto-discovered devices; in fact, I'll even hit the `ignore all` button，And sometimes the public device discovery function cannot complete all device discovery well.

Can I keep it in my own integration?

Reply:
No, you should use the built-in SSDP flow for this. You can set a name / description for any discovered device, thus you are able to customize the presentation of your devices.

You can still add a regular config flow where people can add the integration manually (by IP e.g.) if that is possible.

Reply:
@bdraco any advice from you here? I am not really familiar with SSDP discovery and what we support.

Custom SSDP code from underlying library:
https://github.com/altersis/aioleviosa/blob/6d09ce3f42decd9012246bad75afafd0fdb31fb0/aioleviosa/aioleviosa.py#L29-L71

Reply:
@greg-ha-1990 can you try to better explain what problem is caused if you use the built-in discovery?
Is the problem that if the user has 8 hubs, they would see something like this (although obviously for leviosa, not for shelly):
![image](https://github.com/home-assistant/core/assets/14281572/d472cca0-89e7-4578-a075-f0823b824a56)

In a previous reply, you mentioned you want to present the discovered hubs in a way which is helpful to the user, but in the code it seems like you just generate a list of IP-addresses?

Reply:
> can you try to better explain what problem is caused if you use the built-in discovery?

There is no way to automatically discover curtains on the Devices & Services homepage. This is the current problem.

> In a previous reply, you mentioned you want to present the discovered hubs in a way which is helpful to the user, but in the code it seems like you just generate a list of IP-addresses?

Of course the user experience is not good. First of all, on the Devices & Services homepage, automatically discovered devices always pop up and prompt the user to configure them. This process is completely automated in the background and discovered silently. There is no progress bar or prompt. My curtains , has not been found, I don't know why. This is the so-called poor user experience, with no errors and no prompts. I'm tired of automatic discovery, so I added a "search process" pop-up window to my integration, with a 20-second timeout. If no search is found, an error will be prompted. This is a user-friendly design. As for what you said Only the IP is displayed, because in this scenario, the only thing that needs to be concerned about is the IP, the so-called MAC, and the serial number. Not to mention that only the IP and URN can be found through SSDP. Even if the manufacturer encapsulates the MAC in the URN, then what? What's the point? The IP of the curtain has changed, but the MAC remains unchanged, and it still cannot be controlled. We must understand that we are working in a LAN scenario, and we must accurately know the IP address of the curtain in order to control it.

Reply:
> We want to avoid running multiple SSDP listeners as we already have a performance issue with processing traffic volume on some systems.
> 
> Another option is to register for a callback. Example:
> 
> https://github.com/home-assistant/core/blob/017b2623e8ff741276902578f32aa896c89eed44/homeassistant/components/sonos/__init__.py#L583

Thank you for your reply. The main problem I have is that I don't find these curtains on the homepage. They don't appear like they appear in your screenshots? I would like to add a progress bar with a 20 second timeout to prompt the user for an error. Additionally, the SSDP discovery process will only be enabled during configuration, so the efficiency issue may not be a bottleneck. In order to add a 20 second timeout error message, can I use SSDP discovery in my integration?